sdk/core/Azure.Core/src/TelemetryDetails.cs

```diff
@@ -2,47 +2,60 @@
 // Licensed under the MIT License.
 
 using System;
+using System.ComponentModel;
 using System.Reflection;
 using System.Runtime.InteropServices;
+using Azure.Core.Pipeline;
 
-namespace Azure.Core.Pipeline
+namespace Azure.Core
 {
     /// <summary>
-    /// Information about the package to be included in UserAgent telemetry
+    /// Details about the package to be included in UserAgent telemetry
     /// </summary>
-    public class UserAgentValue
+    public class TelemetryDetails
     {
-        private string _userAgent;
+        private readonly string _userAgent;
 
         /// <summary>
-        /// Initialize an instance of <see cref="UserAgentValue"/> by extracting the name and version information from the <see cref="Assembly"/> associated with the <paramref name="type"/>.
+        /// The package type represented by this <see cref="TelemetryDetails"/> instance.
         /// </summary>
-        /// <param name="type">The <see cref="Type"/> used to generate the package name and version information for the <see cref="UserAgentValue"/> value.</param>
-        /// <param name="applicationId">An optional value to be prepended to the <see cref="UserAgentValue"/>.
+        public Assembly Assembly { get; }
+
+        /// <summary>
+        /// The value of the applicationId used to initialize this <see cref="TelemetryDetails"/> instance.
+        /// </summary>
+        public string? ApplicationId { get; }
+
+        /// <summary>
+        /// The properly formatted UserAgent string based on this <see cref="TelemetryDetails"/> instance.
+        /// </summary>
+        public string UserAgent => _userAgent;
+
+        /// <summary>
+        /// Initialize an instance of <see cref="TelemetryDetails"/> by extracting the name and version information from the <see cref="System.Reflection.Assembly"/> associated with the <paramref name="assembly"/>.
+        /// </summary>
+        /// <param name="assembly">The <see cref="System.Reflection.Assembly"/> used to generate the package name and version information for the <see cref="TelemetryDetails"/> value.</param>
+        /// <param name="applicationId">An optional value to be prepended to the <see cref="TelemetryDetails"/>.
         /// This value overrides the behavior of the <see cref="DiagnosticsOptions.ApplicationId"/> property for the <see cref="HttpMessage"/> it is applied to.</param>
-        public UserAgentValue(Type type, string? applicationId = null)
+        public TelemetryDetails(Assembly assembly, string? applicationId = null)
         {
-            var assembly = Assembly.GetAssembly(type);
-            if (assembly == null) throw new ArgumentException($"The type parameter {type.FullName} does not have a valid Assembly");
+            Argument.AssertNotNull(assembly, nameof(assembly));
+
+            Assembly = assembly;
+            ApplicationId = applicationId;
```

Comment:
Should we be validating the length a la https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/DiagnosticsOptions.cs#L113?

Reply:
good catch.